trantor/net/TcpConnection.h

```diff
@@ -58,20 +56,49 @@ class TcpConnection
     virtual void shutdown() = 0;
     virtual void forceClose() = 0;
     virtual EventLoop *getLoop() = 0;
+    template <typename T>
+    void setContext(T &&context)
+    {
+        typedef typename std::remove_cv<
+            typename std::remove_reference<T>::type>::type ContextType;
+        _contextPtr = std::make_shared<ContextType>(std::forward<T>(context));
+    }
+    template <int N>
+    void setContext(const char (&str)[N])
+    {
+        _contextPtr = std::make_shared<const char *>(str);
+    }
+    template <typename T>
+    const T &getContext() const
+    {
+        assert(_contextPtr);
+        return (*(const T *)_contextPtr.get());
+    }
 
-    virtual void setContext(const any &context) = 0;
-    virtual void setContext(any &&context) = 0;
-    virtual const any &getContext() const = 0;
-
-    virtual any *getMutableContext() = 0;
-
+    template <typename T>
+    T &getContext()
+    {
+        assert(_contextPtr);
+        return (*(T *)_contextPtr.get());
+    }
+    bool hasContext()
+    {
+        return (bool)_contextPtr;
```

Comment:
Casting to bool even via `operator bool` is not always a safe action as a shared_ptr could potentially own `null` memory, thus returning `true` even when the internal pointer is `nullptr`.  Especially if you have a potential reference access but in any case you want to both test if the shared_ptr is empty and also if its internal pointer is not `nullptr`.